docs/2019-12-16-Question-Mark-Variables-With-Ensures.md

```diff
@@ -0,0 +1,313 @@
+
+#### Rand-like examples
+
+1. `randBounded(M,N)` can rewrite to any integer between `M` and `N`
+    ```
+    \forall M:Int.\forall N:Int.\forall I:Int.
+        I >= M /\ I <= N -> randBounded(M, N) -> * I
+    ```
+
+1. `randInList(Is)` takes a list `Is` of items
+   and can rewrite in one step to any item in `Is`.
+
+    ```
+    \forall Is:List . \forall I . I \in Is -> randInList(Is) -> * I
+    ```
+
+1. `randNotInList(Is)` takes a list `Is` of items
+   and can rewrite in one step to any item _not_ in `Is`.
+
+    ```
+    \forall Is:List . \forall I . \not(I \in Is) -> randNotInList(Is) -> * I
+    ```
+
+1. `randPrime()`, can rewrite to any _prime number_.
+
+    ```
+    \forall P:Int . isPrime(P) -> randPrime() -> * P
+    ```
+
+   where `isPrime(_)` is a predicate that can be defined in the usual way.
+
+### Frontend Syntax
+
+Since all variables are universally quantified, `rand()` and its variants could
+be written in the K frontend using regular rules and variables:
+
+```
+syntax Exp ::= rand ()
+rule rand() => X:Int
+
+syntax Exp ::= randBounded(Int, Int)
+rule randBounded(M, N) => I
+  requires M <=Int I andBool I <=Int N
+
+syntax Exp ::= randInList (List)
+rule randInList(Is) => I
+  requires I inList Is
+
+syntax Exp ::= randNotInList (List)
+rule randNotInList(Is) => I
+  requires notBool(I inList Is)
+
+syntax Exp ::= randPrime ()
+rule randPrime() => X:Int
+  requires isPrime(X)
+```
+
+Note 1: all above are not function symbols, but language constructs.
+
+Note 2: Currently the frontend does not allow rules with universally quantified
+variables in the RHS which are not bound in the LHS.
+
+Note 3. Allowing these rules in a concrete execution engine would require an
+algorithm for generating concrete instances for such variables, satisfying the
+given constraints.
+Hence, it might be useful to mark these rules with a specific keyword to signal
+this fact.
+
+## Example B: Fresh Integer Construct `fresh(Is)`
+
+The fresh integer construct `fresh(Is)` is a language construct and part of the
+PL syntax.
+
+```
+Exp ::= ... | "fresh" "(" List{Int} ")"
+```
+
+### Axioms
+
+The intended semantics of `fresh(Is)` is that it can always rewrite to an integer
+that in not in `Is`.
+
+Note that `fresh(Is)` and `randNotInList(Is)` are different; the former
+does not _need_ to be able to rewrite to every integers not in `Is`,
+while the latter requires so.
+
+For example, it is _correct_ to implement `fresh(Is)` so it always returns the
+smallest positive integer that is not in `Is`, but same implementation for
+`randNotInList(Is)` might be considered _inadequate_.
+In other words, there exist multiple correct implementations of `fresh(Is)`,
+some of which may be deterministic, but there only exists a unique
+implementation of `randNotInList(Is)`.
+Finally, note that `randNotInList(Is)` is a _correct_ implementation
+for `fresh(Is)`.
+
+Here is the axiom of `fresh(Is)`
+
+```
+(FRESH) \forall Is:List{Int} . \exists I:Int .
```

Comment:
I'm confused by this, because it says above that the `?` variables will be quantified over the right-hand side only.

Reply:
My bad. I understood Xiaohong's suggestion as adding text, while he obviously meant replacing.  The idea is that we sometimes can move the quantifier to the RHS, but other times, like in this case, we cannot.